Viewport layout size is sometimes incorrect when a site specifies an initial scale

rr-codes opened the pull request:
#### 3cfe137626e4ea11429127409326b27a3726fc58
<pre>
Viewport layout size is sometimes incorrect when a site specifies an initial scale
<a href="https://bugs.webkit.org/show_bug.cgi?id=257583">https://bugs.webkit.org/show_bug.cgi?id=257583</a>
rdar://109483331

Reviewed by Dean Jackson.

On iOS, if a page specifies a viewport meta tag including an initial scale value that is &lt; 1,
and if the `minimumEffectiveDeviceWidth` is non-zero, the resulting layout width will have an
undesirable value. Instead, the initial scale should be ignored when the `minimumEffectiveDeviceWidth`
is specified when computing the layout width.

This PR fixes this by ignoring the initial scale only when the `minimumEffectiveDeviceWidth` actually
has an effect on the viewport layout width.

Also refactors the `applyViewportArgument` to be simpler.

* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::applyViewportArgument):
(WebCore::ViewportConfiguration::updateConfiguration):
(WebCore::ViewportConfiguration::layoutWidth const):
* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::Parameters::operator== const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/Viewport.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/264842@main">https://commits.webkit.org/264842@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52695d38e351d8902f730d7b7a7fb51f9a6ff48c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8799 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9087 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10452 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8787 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11074 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9057 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11646 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8945 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9950 "2 failures") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10610 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7247 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8045 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15548 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8357 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8194 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11528 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8679 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7078 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7940 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2139 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12152 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8426 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

grorg reviewed (approved):
r=me but I agree with Simon that there should be a test.

aprotyas reviewed (approved):
non-reviewer r+

rr-codes committed:
Viewport layout size is sometimes incorrect when a site specifies an initial scale
https://bugs.webkit.org/show_bug.cgi?id=257583
rdar://109483331

Reviewed by Dean Jackson.

On iOS, if a page specifies a viewport meta tag including an initial scale value that is < 1,
and if the `minimumEffectiveDeviceWidth` is non-zero, the resulting layout width will have an
undesirable value. Instead, the initial scale should be ignored when the `minimumEffectiveDeviceWidth`
is specified when computing the layout width.

This PR fixes this by ignoring the initial scale only when the `minimumEffectiveDeviceWidth` actually
has an effect on the viewport layout width.

Also refactors the `applyViewportArgument` to be simpler.

* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::applyViewportArgument):
(WebCore::ViewportConfiguration::updateConfiguration):
(WebCore::ViewportConfiguration::layoutWidth const):
* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::Parameters::operator== const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/Viewport.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/264842@main
